frontend/src/metabase/query_builder/actions/core/initializeQB.ts

```diff
@@ -0,0 +1,337 @@
+  card.original_card_id = card.id;
+
+  return { card, originalCard };
+}
+
+async function fetchAndPrepareAdHocQuestionCards(deserializedCard: Card) {
+  if (!deserializedCard.original_card_id) {
+    return {
+      card: deserializedCard,
+      originalCard: null,
+    };
+  }
+
+  const originalCard = await loadCard(deserializedCard.original_card_id);
+
+  if (cardIsEquivalent(deserializedCard, originalCard)) {
+    return {
+      card: { ...originalCard },
+      originalCard: originalCard,
+    };
+  }
+
+  return {
+    card: deserializedCard,
+    originalCard,
+  };
+}
+
+type ResolveCardsResult = {
+  card: Card;
+  originalCard?: Card;
+};
+
+async function resolveCards({
+  cardId,
+  deserializedCard,
+  options,
+}: {
+  cardId?: number;
+  deserializedCard?: Card;
+  options: BlankQueryOptions;
+}): Promise<ResolveCardsResult> {
+  if (!cardId && !deserializedCard) {
+    return {
+      card: getCardForBlankQuestion(options),
+    };
+  }
+  return cardId
+    ? fetchAndPrepareSavedQuestionCards(cardId)
+    : fetchAndPrepareAdHocQuestionCards(deserializedCard as Card);
+}
+
+function parseHash(hash?: string) {
+  let options: BlankQueryOptions = {};
+  let serializedCard;
+
+  // hash can contain either query params starting with ? or a base64 serialized card
+  if (hash) {
+    const cleanHash = hash.replace(/^#/, "");
+    if (cleanHash.charAt(0) === "?") {
+      options = querystring.parse(cleanHash.substring(1));
+    } else {
+      serializedCard = cleanHash;
+    }
+  }
+
+  return { options, serializedCard };
+}
+
+function isSavedCard(card: Card): card is SavedCard {
+  return !!(card as SavedCard).id;
+}
+
+export const INITIALIZE_QB = "metabase/qb/INITIALIZE_QB";
+
+async function handleQBInit(
+  dispatch: Dispatch,
+  getState: GetState,
+  {
+    location,
+    params,
+  }: { location: LocationDescriptorObject; params: QueryParams },
+) {
+  dispatch(resetQB());
+  dispatch(cancelQuery());
+
+  const queryParams = location.query;
+  const cardId = Urls.extractEntityId(params.slug);
+  const uiControls: UIControls = getQueryBuilderModeFromLocation(location);
+  const { options, serializedCard } = parseHash(location.hash);
+  const hasCard = cardId || serializedCard;
+
+  if (
+    !hasCard &&
+    !options.db &&
+    !options.table &&
+    !options.segment &&
+    !options.metric
+  ) {
+    dispatch(redirectToNewQuestionFlow());
+    return;
+  }
+
+  const deserializedCard = serializedCard
+    ? deserializeCard(serializedCard)
+    : null;
+
+  const { card, originalCard } = await resolveCards({
+    cardId,
+    deserializedCard,
+    options,
+  });
+
+  if (isSavedCard(card) && card.archived) {
+    dispatch(setErrorPage(ARCHIVED_ERROR));
+    return;
+  }
+
+  if (
+    isSavedCard(card) &&
+    !card?.dataset &&
+    location.pathname?.startsWith("/model")
+  ) {
+    dispatch(setErrorPage(NOT_FOUND_ERROR));
+    return;
+  }
```

Comment:
Before the clean-up, `initializeQB` was not exiting after these checks and was loading query metadata, snippets, etc. without any clear reason. Please LMK if it looks wrong to you to exit the function at this point